![Callysto.ca Banner](https://github.com/callysto/curriculum-notebooks/blob/master/callysto-notebook-banner-top.jpg?raw=true)

<a href="https://hub.callysto.ca/jupyter/hub/user-redirect/git-pull?repo=https%3A%2F%2Fgithub.com%2Fcallysto%2Fcurriculum-notebooks&branch=master&subPath=SocialStudies/HansardAnalysis/hansard-analysis.ipynb&depth=1" target="_parent"><img src="https://raw.githubusercontent.com/callysto/curriculum-notebooks/master/open-in-callysto-button.svg?sanitize=true" width="123" height="24" alt="Open in Callysto"/></a>

# Hansard Analysis

The [Hansard](https://en.wikipedia.org/wiki/Hansard) is a transcript of debates in the Canadian Parliament. It is available from the official [Parliament of Canada website](https://www.parl.ca) as well as other sources such as [Open Parliament](https://openparliament.ca) and [LiPaD: The Linked Parliamentary Data Project](https://www.lipad.ca).

We have downloaded the 2020 files from LiPaD, and can load them by selecting the following code cell and clicking the `▶Run` button.

In [1]:
import pandas as pd
hansard = pd.read_csv('proceedings2020.csv')
print(hansard.shape)
hansard.columns

(4945, 15)


Index(['basepk', 'hid', 'speechdate', 'pid', 'opid', 'speakeroldname',
       'speakerposition', 'maintopic', 'subtopic', 'subsubtopic', 'speechtext',
       'speakerparty', 'speakerriding', 'speakername', 'speakerurl'],
      dtype='object')

There are 15 columns and 4945 rows of data.

## Who Spoke?

Let's have a look at who spoke during these debates.

In [ ]:
speakers = hansard.drop_duplicates(subset=['speakername'])[['speakername','speakerparty','speakerriding','speakerurl']]
speakers = speakers.dropna().reset_index().drop(columns=['index'])
print('There were',speakers.shape[0],'speakers from the',speakers['speakerparty'].unique(),'parties.')

We can compare that to the list of Members of Parliament from the [43rd Parliament](https://en.wikipedia.org/wiki/43rd_Canadian_Parliament) that started on December 5, 2019.

In [ ]:
members = pd.read_csv('https://www.ourcommons.ca/members/en/search/csv?parliament=43')
print('There were',members.shape[0],'Members from the',members['Political Affiliation'].unique(),'parties.')

So of the 340 Members of Parliament we had 312 unique speakers, meaning that 28 Members are not recorded as speaking during 2020. Let's see if we can identify who are they were.

In [ ]:
members['Name'] = members['First Name'] +' '+ members['Last Name']
silent = []
for member in members['Name']:
    if member not in speakers['speakername'].values:
        silent.append(member)
print('That is',len(silent),'Members not recorded as speaking in 2020:')
print(silent)

Of course 35 is not equal to 28, but we will leave it to you to compare the list `silent` to the list from `speakers['speakername'].unique()` if you are interested.

## Who Spoke Most?

We can check how many times each speaker is recorded in the Hansard.

In [ ]:
hansard['speakername'].value_counts()

Let's also calculate the length (number of characters) of each of those speeches, and sort them by who said the most.

In [ ]:
hansard['speechlength'] = hansard['speechtext'].str.len()
hansard.groupby('speakername').sum().sort_values('speechlength', ascending=False)

We can also plot those speech lengths. Feel free to change the `n = 20` variable to see more or fewer Members.

In [ ]:
import plotly.express as px
n = 20
top_length = hansard.groupby('speakername').sum().sort_values('speechlength', ascending=False).head(n)
px.bar(top_length, title='Top '+str(n)+' Hansard Speakers by Speech Length').update(layout_showlegend=False)

As well, we can check how many times a Member from a political party spoke.

In [ ]:
px.bar(hansard['speakerparty'].value_counts(),title='Hansard Speaker Frequency by Party').update(layout_showlegend=False)

In [ ]:
speech_length_party = hansard.groupby('speakerparty').sum().sort_values('speechlength', ascending=False)
px.bar(speech_length_party, title='Hansard Speech Length by Party').update(layout_showlegend=False)

## Natural Language Processing

We are going to use the Python library [spaCy](https://spacy.io) for natural language processing. It will allow us to remove "stop words", which are common words that can be discarded without reducing meaning. We will also simplify words to their base forms, which is called "[lemmatization](https://en.wikipedia.org/wiki/Lemmatisation)" (e.g. the lemma of "speaking" is "speak").

The following code cell will take a while to run.

In [15]:
try:
    import spacy
    nlp = spacy.load("en_core_web_sm")
except:
    !pip install spacy --user
    !python -m spacy download en_core_web_sm
    import spacy
    nlp = spacy.load("en_core_web_sm")
from spacy.lang.en.stop_words import STOP_WORDS as sw
sw.add('madam')
sw.add('speaker')
sw.add('hon')
sw.add('member')
sw.add('\n')
sw.add('mr.')
sw.add('government')
sw.add('canada')
sw.add('prime')
sw.add('minister')

def tokenize(text):
    speech = []
    try:
        for token in nlp(text):
            if token.is_stop != True and token.is_punct != True and token != '\n':
                speech.append(token.lemma_)
    except:
        pass
    return speech

hansard['tokenized'] = hansard['speechtext'].apply(tokenize)
hansard

basepk                            hid  speechdate  \
0     4786557  ca.proc.d.2020-01-27.10733547  2020-01-27   
1     4786575  ca.proc.d.2020-01-27.10733601  2020-01-27   
2     4786602  ca.proc.d.2020-01-27.10733669  2020-01-27   
3     4786603  ca.proc.d.2020-01-27.10733670  2020-01-27   
4     4786535  ca.proc.d.2020-01-27.10733478  2020-01-27   
5     4786536  ca.proc.d.2020-01-27.p6026577  2020-01-27   
6     4786537  ca.proc.d.2020-01-27.10733485  2020-01-27   
7     4786538  ca.proc.d.2020-01-27.10733492  2020-01-27   
8     4786539  ca.proc.d.2020-01-27.10733494  2020-01-27   
9     4786555  ca.proc.d.2020-01-27.10733540  2020-01-27   
10    4786541  ca.proc.d.2020-01-27.10733499  2020-01-27   
11    4786543  ca.proc.d.2020-01-27.10733503  2020-01-27   
12    4786544  ca.proc.d.2020-01-27.10733505  2020-01-27   
13    4786546  ca.proc.d.2020-01-27.10733517  2020-01-27   
14    4786547  ca.proc.d.2020-01-27.10733519  2020-01-27   
15    4786548  ca.proc.d.2020-01-27.10733521  2020-01-27   
16    4786550  ca.proc.d.2020-01-27.10733526  2020-01-27   
17    4786551  ca.proc.d.2020-01-27.10733528  2020-01-27   
18    4786552  ca.proc.d.2020-01-27.10733534  2020-01-27   
19    4786553  ca.proc.d.2020-01-27.10733536  2020-01-27   
20    4786559  ca.proc.d.2020-01-27.10733552  2020-01-27   
21    4786562  ca.proc.d.2020-01-27.10733563  2020-01-27   
22    4786564  ca.proc.d.2020-01-27.10733568  2020-01-27   
23    4786561  ca.proc.d.2020-01-27.10733557  2020-01-27   
24    4786570  ca.proc.d.2020-01-27.10733582  2020-01-27   
25    4786571  ca.proc.d.2020-01-27.10733592  2020-01-27   
26    4786572  ca.proc.d.2020-01-27.10733595  2020-01-27   
27    4786574  ca.proc.d.2020-01-27.10733600  2020-01-27   
28    4786576  ca.proc.d.2020-01-27.10733602  2020-01-27   
29    4786578  ca.proc.d.2020-01-27.10733606  2020-01-27   
...       ...                            ...         ...   
4915  4791337  ca.proc.d.2020-12-10.11082099  2020-12-10   
4916  4791339  ca.proc.d.2020-12-10.11082105  2020-12-10   
4917  4791384  ca.proc.d.2020-12-10.11082231  2020-12-10   
4918  4791392  ca.proc.d.2020-12-10.11082283  2020-12-10   
4919  4791412  ca.proc.d.2020-12-10.11082480  2020-12-10   
4920  4791421  ca.proc.d.2020-12-10.11082557  2020-12-10   
4921  4791427  ca.proc.d.2020-12-10.11082587  2020-12-10   
4922  4791433  ca.proc.d.2020-12-10.11082760  2020-12-10   
4923  4791434  ca.proc.d.2020-12-10.11082765  2020-12-10   
4924  4791442  ca.proc.d.2020-12-10.11082944  2020-12-10   
4925  4791459  ca.proc.d.2020-12-10.11083627  2020-12-10   
4926  4791176  ca.proc.d.2020-12-10.11080402  2020-12-10   
4927  4791470  ca.proc.d.2020-12-11.p6431686  2020-12-11   
4928  4791471  ca.proc.d.2020-12-11.11084303  2020-12-11   
4929  4791472  ca.proc.d.2020-12-11.11084327  2020-12-11   
4930  4791473  ca.proc.d.2020-12-11.11084332  2020-12-11   
4931  4791475  ca.proc.d.2020-12-11.11084319  2020-12-11   
4932  4791474  ca.proc.d.2020-12-11.11084315  2020-12-11   
4933  4791476  ca.proc.d.2020-12-11.11084321  2020-12-11   
4934  4791477  ca.proc.d.2020-12-11.11084333  2020-12-11   
4935  4791478  ca.proc.d.2020-12-11.11084326  2020-12-11   
4936  4791479  ca.proc.d.2020-12-11.11084335  2020-12-11   
4937  4791480  ca.proc.d.2020-12-11.11084337  2020-12-11   
4938  4791482  ca.proc.d.2020-12-11.11084342  2020-12-11   
4939  4791484  ca.proc.d.2020-12-11.11084349  2020-12-11   
4940  4791485  ca.proc.d.2020-12-11.11084351  2020-12-11   
4941  4791486  ca.proc.d.2020-12-11.11084376  2020-12-11   
4942  4791487  ca.proc.d.2020-12-11.11084360  2020-12-11   
4943  4791481  ca.proc.d.2020-12-11.11084339  2020-12-11   
4944  4791483  ca.proc.d.2020-12-11.11084347  2020-12-11   

                                       pid    opid  \
0     ddcc96da-5bcf-4069-9cd6-da8f76df8b89  229530   
1     cedf1659-ebb1-4f26-b723-5c7c5353988b  214948   
2     cedf1659-ebb1-4f26-b723-5c7c5353988b  214948   
3     00000000-0000-0000-0000-000000000000  248690   
4     3dd9162a-4

In [16]:
from collections import Counter
for speaker in hansard['speakername'].unique():
    words = []
    for token_list in hansard[hansard['speakername']==speaker]['tokenized']:
        words.extend(token_list)
    word_freq = Counter(words)
    common_words = word_freq.most_common(5)
    print(speaker, common_words)

Stéphane Lauzon []
Carol Hughes []
Paul Manly []
Chrystia Freeland []
nan []
Raquel Dancho []
Kevin Lamoureux []
Greg Fergus []
John Brassard []
Dan Albas []
Alexandre Boulerice []
Marie-France Lalonde []
Darrell Samson []
Jacques Gourde []
Wayne Easter []
Jeremy Patzer []
Damien Kurek []
Elizabeth May []
Garnett Genuis []
Tom Kmiec []
Julie Dzerowicz []
Bryan May []
Earl Dreeshen []
Kenneth McDonald []
Anthony Housefather []
Marc Dalton []
William Amos []
Jean Yip []
James Cumming []
Majid Jowhari []
Michael Cooper []
Nelly Shin []
Alistair MacGregor []
Michel Boudrias []
Mario Simard []
Brian Masse []
Yves Perron []
Alexis Brunelle-Duceppe []
Denis Trudel []
Martin Champoux []
Heather McPherson []
Lloyd Longfield []
Xavier Barsalou-Duval []
The Speaker []
Some hon. members []
Luc Berthold []
Cathy McLeod []
Peter Julian []
Mark Gerretsen []
Andy Fillmore []
Gérard Deltell []
Brad Vis []
Charlie Angus []
Robert Gordon Kitchen []
Gord Johns []
Gabriel Ste-Marie []
Anthony Rota []
Micha

## Other Datasets

If you want to explore datasets from other timeframes on [LiPaD](https://www.lipad.ca/full/), you can upload the CSV files to a `new_data` folder here, and run the following code to import them into a DataFrame.

```python
import os
import pandas as pd
hansard = pd.DataFrame()
for root, dirs, files in os.walk('new_data'):
    for name in files:
        df = pd.read_csv(os.path.join(root, name))
        hansard = hansard.append(df)
hansard
```

# Conclusion



[![Callysto.ca License](https://github.com/callysto/curriculum-notebooks/blob/master/callysto-notebook-banner-bottom.jpg?raw=true)](https://github.com/callysto/curriculum-notebooks/blob/master/LICENSE.md)